### PR TITLE
fix: GCS data source returns part stream failure if source blob doesn't exist

### LINKED ISSUE #80: Avoid creating an empty blob when transferring an asset referring to non-existent blob with the GCS extension
# Bug Report

## Describe the Bug
When transferring an asset referring to non-existent blob, an empty blob is created on the consumer side.

### Expected Behavior
In case the transfer request is about an asset referring to a non-existing blob, report the exception and avoid creating an empty blob.

## Context Information
EDC v 0.3.2-SNAPSHOT

## Possible Implementation
Throw a runtime exception when invoking the public InputStream openStream() method of a Part referring to a non-existing blob.

### PR DESCRIPTION
**Pull request by man8pr**
- GcsDataSource.GoogleStoragePart openStream throws EdcException if the requested blob doesn't exist
- GcsDataSink transferParts handles EdcException, plus minor update in log messages

## Linked Issue(s)

Closes #80

### CODE REVIEW

**Review by ndr-brt — CHANGES_REQUESTED** [review]
I think it would be better that if the source blob does not exist the `Part` object should not be created, so the validation should be done in the `DataSource.openPartStream` method.
In any case, please add tests

**Comment by man8pr** [comment]
> I think it would be better that if the source blob does not exist the `Part` object should not be created, so the validation should be done in the `DataSource.openPartStream` method. In any case, please add tests

Good point, I was trying to have the blob existence test at the latest point in time, what do you think?

**Comment by ndr-brt** [comment]
> > I think it would be better that if the source blob does not exist the `Part` object should not be created, so the validation should be done in the `DataSource.openPartStream` method. In any case, please add tests
> 
> Good point, I was trying to have the blob existence test at the latest point in time, what do you think?

I think the amount of time that passes from the `openPartStream` to the `openStream` is negligible.
We did the same for the [Blobstorage data source](https://github.com/eclipse-edc/Technology-Azure/blob/main/extensions/data-plane/data-plane-azure-storage/src/main/java/org/eclipse/edc/connector/dataplane/azure/storage/pipeline/AzureStorageDataSource.java) and the [S3 one](https://github.com/eclipse-edc/Technology-Aws/blob/main/extensions/data-plane/data-plane-aws-s3/src/main/java/org/eclipse/edc/connector/dataplane/aws/s3/S3DataSource.java)

File: extensions/data-plane/data-plane-google-storage/src/main/java/org/eclipse/edc/connector/dataplane/gcp/storage/GcsDataSink.java [hunk 1]
```diff
@@ -52,10 +53,14 @@ protected StreamResult<Object> transferParts(List<DataSource.Part> parts) {
             } catch (IOException e) {
                 monitor.severe("Cannot open the input part", e);
                 monitor.severe(e.toString());
-                return StreamResult.error("An error");
+                return StreamResult.error("Cannot open the input part");
+            } catch (EdcException e) {
+                monitor.severe("Cannot open the source blob", e);
+                monitor.severe(e.toString());
+                return StreamResult.error("Cannot open the source blob");
```
Comment (ndr-brt) [inline extensions/data-plane/data-plane-google-storage/src/main/java/org/eclipse/edc/connector/dataplane/gcp/storage/GcsDataSink.java:60]: could this catch branch need to be tested

File: extensions/data-plane/data-plane-google-storage/src/test/java/org/eclipse/edc/connector/dataplane/gcp/storage/GcsDataSourceTest.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+import org.eclipse.edc.spi.monitor.Monitor;
+import org.junit.jupiter.api.Test;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+public class GcsDataSourceTest {
+
+    Monitor monitor = mock(Monitor.class);
```
Comment (ndr-brt) [inline extensions/data-plane/data-plane-google-storage/src/test/java/org/eclipse/edc/connector/dataplane/gcp/storage/GcsDataSourceTest.java:29]: nit: is not necessary to pass the class type to the `mock` method when the type is explicit

File: extensions/data-plane/data-plane-google-storage/src/main/java/org/eclipse/edc/connector/dataplane/gcp/storage/GcsDataSink.java (hunk 1 above)
Comment (man8pr) [reply extensions/data-plane/data-plane-google-storage/src/main/java/org/eclipse/edc/connector/dataplane/gcp/storage/GcsDataSink.java:60]: The catch block was added to handle the blob existence check in the sink (now removed), not used anymore: removing.

Comment (man8pr) [reply extensions/data-plane/data-plane-google-storage/src/main/java/org/eclipse/edc/connector/dataplane/gcp/storage/GcsDataSink.java]: Removed

File: extensions/data-plane/data-plane-google-storage/src/test/java/org/eclipse/edc/connector/dataplane/gcp/storage/GcsDataSourceTest.java (hunk 2 above)
Comment (man8pr) [reply extensions/data-plane/data-plane-google-storage/src/test/java/org/eclipse/edc/connector/dataplane/gcp/storage/GcsDataSourceTest.java]: Fixed

**Review by ndr-brt — APPROVED** [review]

**Comment by ndr-brt** [comment]
please fix the dependencies file then we'll be ready to go

**Pull request merged**